### PR TITLE
Add metadata to multicast group entry

### PR DESCRIPTION
**Pull request by verios-google**
We have a use case for this. The change is small, backwards compatible, and consistent with table entries.

### CODE REVIEW

**Comment by chrispsommers** [comment]
Looks like you have some trailing whitespace in the .mdk and you need to regenerate the protobuf code, see instructions at bottom of this section: https://github.com/p4lang/p4runtime#detailed-processes. Thanks.

**Review by smolkaj — COMMENTED** [review]
LGTM from my side. 

Let's also give others a chance to weigh in, e.g. @saynb @antoninbas @jafingerhut.

**Review by jafingerhut — APPROVED** [review]
LGTM

**Comment by jafingerhut** [comment]
@verios-google @smolkaj Do you use Stratum P4Runtime API implementation code in your work?  If yes, am I correct in guessing that implementing these changes would require some (probably small) changes in Stratum to accomplish this?  Are you willing to do that?

**Comment by jafingerhut** [comment]
Same question for whether these changes might require implementation code changes in this repo: https://github.com/p4lang/PI

File: proto/p4/v1/p4runtime.proto [hunk 1]
```diff
@@ -451,6 +451,8 @@ message Replica {
 message MulticastGroupEntry {
   uint32 multicast_group_id = 1;
   repeated Replica replicas = 2;
+  // Arbitrary metadata from the controller that is opaque to the target.
```
Comment (jafingerhut) [inline proto/p4/v1/p4runtime.proto:454]: There should be a comment about which version of the P4Runtime API specification in which this field was added.  See https://github.com/p4lang/p4runtime/issues/441

**Comment by jafingerhut** [comment]
@saynb Are you the right person to ask about Stratum developer contacts?

@smolkaj I think it is perfectly reasonable to ask Antonin about development effort required to make such a change in p4lang/PI, but I would recommend that serious thought be given to the idea that someone besides Antonin actually implements it.  Otherwise, the base of developers who know how to make such changes will not be growing.

**Comment by antoninbas** [comment]
> We do use BMv2, so extending https://github.com/p4lang/PI seems like a reasonable ask. @antoninbas any ideas how much work this may be?

This is very simple to implement. Should be about the same scope as your recent PR (https://github.com/p4lang/PI/pull/598).

The `metadata` binary string can probably be added directly to the existing `Group` struct:
https://github.com/p4lang/PI/blob/e86801b69a51fc3462ed67ac2f7fa6762e827460/proto/frontend/src/pre_mc_mgr.h#L92

**Comment by smolkaj** [comment]
> but I would recommend that serious thought be given to the idea that someone besides Antonin actually implements it. Otherwise, the base of developers who know how to make such changes will not be growing.

Sure, I agree. I imlemented https://github.com/p4lang/PI/pull/598 just a few days ago.

**Comment by saynb** [comment]
> @saynb Are you the right person to ask about Stratum developer contacts?

I am not the primary person but I can help connect the right people in addition to me.
[ipdk-stratum](https://github.com/ipdk-io/stratum-dev/tree/1c3038100f1ebe01292d3e35b397f12815c1516a) : @ffoulkes @vsureshkumarp would be the contacts.

[ONF stratum](https://github.com/stratum/stratum) : Probably @pudelkoM  ?

**Comment by pudelkoM** [comment]
While most of us don't work at ONF anymore, I can certainly take a look at any potential PR and merge it.

**Review by smolkaj — APPROVED** [review]

**Comment by jafingerhut** [comment]
It appears that automated CI checks for this PR never completed.  I do not know why.

Likely one of those checks that would fail if it had completed is one for checking in the auto-generated output files from the modified .proto files in your PR.  Instructions for regenerating those files can be found near the end of this section of the README: https://github.com/p4lang/p4runtime#detailed-processes

Committing such updated files should trigger CI checks to re-run (or any other commit to this PR).

**Comment by chrispsommers** [comment]
> It appears that automated CI checks for this PR never completed. I do not know why.
> 
> Likely one of those checks that would fail if it had completed is one for checking in the auto-generated output files from the modified .proto files in your PR. Instructions for regenerating those files can be found near the end of this section of the README: https://github.com/p4lang/p4runtime#detailed-processes
> 
> Committing such updated files should trigger CI checks to re-run (or any other commit to this PR).

See my comment https://github.com/p4lang/p4runtime/pull/446#issuecomment-1675433135

**Review by chrispsommers — APPROVED** [review]

File: proto/p4/v1/p4runtime.proto (hunk 1 above)
Comment (chrispsommers) [reply proto/p4/v1/p4runtime.proto:454]: On 9/8/2023 we agreed to incrementally add to the revision history in the main .mdk spec.

**Review by antoninbas — APPROVED** [review]
LGTM

**Pull request merged**